### PR TITLE
Deprecate Test task default classpath convention

### PR DESCRIPTION
**Pull request by jvandort**
Fixes: https://github.com/gradle/gradle/issues/23883

### CODE REVIEW

**Comment by jvandort** [comment]
@bot-gradle test bd prf

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc [hunk 1]
```diff
@@ -100,6 +100,41 @@ intended for internal use and was never intended to be used as part of the publi
...
+.build.gradle
+[source,groovy]
+----
+tasks.withType(Test).configureEach {
+    testClassesDirs = tasks.test.testClassesDirs
+    classpath = tasks.test.classpath
+}
+----
+=====
+====
```
Comment (tresat) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:135]: 💭 Maybe add a brief sentence or two about how this old behavior could lead to confusion when creating custom `Test` tasks which don't need these conventions.

**Review by tresat — APPROVED** [review]

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc (hunk 1 above)
Comment (jvandort) [reply subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:135]: Thanks for the quick review, though I had accidentally marked this as non-draft. This PR still need a bit of work to fix up some of the samples. If there are any changes worth reviewing I'll let you know once I've updated the samples.

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test and merge

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc [hunk 2]
```diff
@@ -100,6 +100,47 @@ intended for internal use and was never intended to be used as part of the publi
 
 Starting in Gradle 9.0, this method will be removed without replacement.
 
+[[test_task_default_classpath]]
+==== Relying on Test task default testClassesDirs and classpath
+
+By default, when applying the `JavaPlugin`, standalone `Test` tasks are configured such that their `testClassesDirs`
```
Comment (big-guy) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:106]: Instead of using the plugin's class name, we should use it's plugin id and link to its user manual page

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc [hunk 3]
```diff
@@ -100,6 +100,47 @@ intended for internal use and was never intended to be used as part of the publi
...
+and `classpath` mirror the corresponding values in the default `test` `TestSuite`. This behavior will be removed
+in Gradle 9.0.
+
+For most use cases, this existing default behavior is unnecessary, since most standalone test
+tasks are used to model separate test suites and thus run a separate set of tests than those
+of the default test task. For more information on how to properly configure a standalone test task
+to compile and run against a separate set of tests, see link:https://docs.gradle.org/current/userguide/java_testing.html#sec:configuring_java_integration_tests[Configuring integration tests].
```
Comment (big-guy) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:113]: This shouldn't use an explicit link to /current/.  We should be using a asciidoc reference e.g., something like: `<<java_testing#sec:configuring_java_integration_tests>>`

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc [hunk 4]
```diff
@@ -100,6 +100,47 @@ intended for internal use and was never intended to be used as part of the publi
...
+----
+=====
+[.multi-language-sample]
+=====
+.build.gradle
+[source,groovy]
+----
+tasks.myTestTask {
+    testClassesDirs = tasks.test.testClassesDirs
+    classpath = tasks.test.classpath
```
Comment (big-guy) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:137]: Instead of pulling from the test task, we should pull from the test suite or reproduce what the convention is doing.

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc [hunk 5]
```diff
@@ -100,6 +100,47 @@ intended for internal use and was never intended to be used as part of the publi
 
 Starting in Gradle 9.0, this method will be removed without replacement.
 
+[[test_task_default_classpath]]
+==== Relying on Test task default testClassesDirs and classpath
+
+By default, when applying the `JavaPlugin`, standalone `Test` tasks are configured such that their `testClassesDirs`
+and `classpath` mirror the corresponding values in the default `test` `TestSuite`. This behavior will be removed
```
Comment (big-guy) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:107]: I would reword this to repeat what the convention is more explicitly. It doesn't have to be exactly this:

By default, when applying the Java plugin, the `testClassesDirs` and `classpath` for all `Test` tasks have the same convention. Unless changed, a `Test` task will use the tests from the default `test` test suite.

Comment (big-guy) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:107]: Can we also link to the test suite docs?

File: subprojects/docs/src/snippets/buildCache/integration-tests/groovy/build.gradle [hunk 6]
```diff
@@ -2,7 +2,14 @@ plugins {
...
 
-tasks.register("integTest", Test)
+sourceSets {
+    integTest
+}
+
+tasks.register("integTest", Test) {
+    classpath = sourceSets.integTest.runtimeClasspath
+    testClassesDirs = sourceSets.integTest.output.classesDirs
+}
```
Comment (big-guy) [inline subprojects/docs/src/snippets/buildCache/integration-tests/groovy/build.gradle:13]: Maybe just create this via test suites?

File: subprojects/docs/src/snippets/testing/test-suite-plugin/kotlin/build.gradle.kts [hunk 7]
```diff
@@ -35,7 +35,7 @@ testing {
             useJUnitJupiter() // <3>
         }
 
-        val integrationTest by registering(JvmTestSuite::class) { // <4>
+        register<JvmTestSuite>("integrationTest") { // <4>
```
Comment (big-guy) [inline subprojects/docs/src/snippets/testing/test-suite-plugin/kotlin/build.gradle.kts:38]: Why did this change?

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestReportIntegrationTest.groovy [hunk 8]
```diff
@@ -106,6 +106,8 @@ public class LoggingTest {
 $junitSetup
 def test = tasks.named('test', Test)
 test.configure {
+    classpath = tasks.test.classpath
```
Comment (big-guy) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestReportIntegrationTest.groovy:109]: Isn't this the same task?

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestReportIntegrationTest.groovy [hunk 9]
```diff
@@ -114,6 +116,8 @@ test.configure {
 }
 
 def superTest = tasks.register('superTest', Test) {
```
Comment (big-guy) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestReportIntegrationTest.groovy:118]: Can these be replaced with test suites?

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestTaskIntegrationTest.groovy [hunk 10]
```diff
@@ -379,11 +379,18 @@ class TestTaskIntegrationTest extends JUnitMultiVersionIntegrationSpec {
...
+            }
+
             dependencies {
-                testImplementation 'junit:junit:${JUnitCoverage.NEWEST}'
+                customTestImplementation 'junit:junit:${JUnitCoverage.NEWEST}'
             }
 
             tasks.create('customTest', Test) {
+                classpath = sourceSets.customTest.runtimeClasspath
+                testClassesDirs = sourceSets.customTest.output.classesDirs
```
Comment (big-guy) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestTaskIntegrationTest.groovy:393]: This breaks the test because we're now not going to find any test classes.  I think we want to move the source into the right path or configure customTest to use test's conventions.

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/JvmTestSuitePlugin.java [hunk 11]
```diff
@@ -73,14 +74,24 @@ public void apply(Project project) {
...
-        // TODO: Deprecate this behavior?
-        // Why would any Test task created need to use the test source set's classes?
         project.getTasks().withType(Test.class).configureEach(test -> {
             // The test task may have already been created but the test sourceSet may not exist yet.
             // So defer looking up the java extension and sourceSet until the convention mapping is resolved.
             // See https://github.com/gradle/gradle/issues/18622
-            test.getConventionMapping().map("testClassesDirs", () -> JvmPluginsHelper.getDefaultTestSuite(project).getSources().getOutput().getClassesDirs());
-            test.getConventionMapping().map("classpath", () -> JvmPluginsHelper.getDefaultTestSuite(project).getSources().getRuntimeClasspath());
+            test.getConventionMapping().map("testClassesDirs", () -> {
+                DeprecationLogger.deprecate("Configuring the test classes dirs by default for standalone Test tasks")
```
Comment (big-guy) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/JvmTestSuitePlugin.java:82]: This is written in a confusing way.  The user isn't configuring anything.  Maybe...

> Relying on the convention for Test.testClassesDirs in custom Test tasks

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/JvmTestSuitePlugin.java [hunk 12]
```diff
@@ -73,14 +74,24 @@ public void apply(Project project) {
...
+                    .willBeRemovedInGradle9()
+                    .withUpgradeGuideSection(8, "test_task_default_classpath")
+                    .nagUser();
+                return JvmPluginsHelper.getDefaultTestSuite(project).getSources().getOutput().getClassesDirs();
+            });
+            test.getConventionMapping().map("classpath", () -> {
+                DeprecationLogger.deprecate("Configuring the test classpath by default for standalone Test tasks")
```
Comment (big-guy) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/JvmTestSuitePlugin.java:89]: Also fix this one to look like the one above

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc [hunk 13]
```diff
@@ -100,6 +100,47 @@ intended for internal use and was never intended to be used as part of the publi
 
 Starting in Gradle 9.0, this method will be removed without replacement.
 
+[[test_task_default_classpath]]
+==== Relying on Test task default testClassesDirs and classpath
```
Comment (big-guy) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:104]: Maybe?
> Relying on conventions for custom Test tasks

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc [hunk 14]
```diff
@@ -100,6 +100,47 @@ intended for internal use and was never intended to be used as part of the publi
...
+By default, when applying the `JavaPlugin`, standalone `Test` tasks are configured such that their `testClassesDirs`
+and `classpath` mirror the corresponding values in the default `test` `TestSuite`. This behavior will be removed
+in Gradle 9.0.
+
+For most use cases, this existing default behavior is unnecessary, since most standalone test
+tasks are used to model separate test suites and thus run a separate set of tests than those
+of the default test task. For more information on how to properly configure a standalone test task
+to compile and run against a separate set of tests, see link:https://docs.gradle.org/current/userguide/java_testing.html#sec:configuring_java_integration_tests[Configuring integration tests].
+
+However, this existing default behavior can be restored for a given
```
Comment (big-guy) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:115]: We might want to say something like "To silence the deprecation warning, you need to explicitly provide the conventions for a custom task"  Or use test suites?

**Review by big-guy — COMMENTED** [review]
@jvandort @tresat this needs to be addressed for 8.1

File: subprojects/docs/src/snippets/buildCache/integration-tests/groovy/build.gradle (hunk 6 above)
Comment (jvandort) [reply subprojects/docs/src/snippets/buildCache/integration-tests/groovy/build.gradle:13]: I was under the impression that we should not use incubating APIs in samples/snippets.

File: subprojects/docs/src/snippets/testing/test-suite-plugin/kotlin/build.gradle.kts (hunk 7 above)
Comment (jvandort) [reply subprojects/docs/src/snippets/testing/test-suite-plugin/kotlin/build.gradle.kts:38]: This was unrelated to these changes. It was failing in `master` but was marked as ignored in CI. The issue was that Kotlin was complaining about the `integrationTest` variable being unused.

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestReportIntegrationTest.groovy (hunk 8 above)
Comment (jvandort) [reply subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestReportIntegrationTest.groovy:109]: Whoops

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestTaskIntegrationTest.groovy (hunk 10 above)
Comment (jvandort) [reply subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestTaskIntegrationTest.groovy:393]: Moved the source files
